File: server/services/urlSafety.js
/**
 * ==========================================================
 * URLSAFETY.JS - SSRF validation (Rev 5 §9.6)
 * ==========================================================
 *
 * "The admin panel accepts arbitrary URLs and fetches them server-side.
 * That is an SSRF risk... Worth unit-testing — pure function, no mocking,
 * and a silently broken validator is exactly the kind of bug nobody
 * notices."
 *
 * This is Node's SANITY CHECK before ever calling Python — cheap, pure,
 * synchronous, no DNS lookups (a lookup would make this impure/async and
 * unmockable-without-network in tests, which is exactly what the spec
 * doesn't want here). It catches the obvious cases: wrong protocol, a
 * literal private/loopback/link-local IP, non-standard ports.
 *
 * It is NOT the authoritative check. Per §9.3: "Python is the only side
 * that sees the redirect chain" — a hostname that resolves fine here can
 * still redirect to an internal IP mid-fetch, and DNS rebinding means the
 * hostname's IP at validation time isn't guaranteed to be its IP at fetch
 * time either. Python's fetch-time check (re-validating the resolved IP on
 * every redirect hop, see python/url_safety.py) is what actually protects
 * the network — this module only stops obviously-bad input early so the
 * admin gets instant feedback instead of waiting for a round trip.
 */

const BLOCKED_HOSTNAMES = new Set([
  "localhost",
  "ip6-localhost",
  "ip6-loopback",
  "metadata.google.internal", // cloud metadata SSRF vector — not in the spec's literal list, but a standard one to block alongside it
]);

// [network, prefixLength] pairs, spec's exact list (§9.6).
const BLOCKED_IPV4_RANGES = [
  ["127.0.0.0", 8],
  ["10.0.0.0", 8],
  ["172.16.0.0", 12],
  ["192.168.0.0", 16],
  ["169.254.0.0", 16],
  ["0.0.0.0", 8],
];

function ipv4ToInt(ip) {
  const parts = ip.split(".");
  if (parts.length !== 4) return null;
  let n = 0;
  for (const part of parts) {
    if (!/^\d{1,3}$/.test(part)) return null;
    const octet = Number(part);
    if (octet < 0 || octet > 255) return null;
    n = (n << 8) | octet;
  }
  return n >>> 0;
}

function isIpv4InRange(ip, network, prefixLength) {
  const ipInt = ipv4ToInt(ip);
  const netInt = ipv4ToInt(network);
  if (ipInt === null || netInt === null) return false;
  const mask = prefixLength === 0 ? 0 : (0xffffffff << (32 - prefixLength)) >>> 0;
  return (ipInt & mask) === (netInt & mask);
}

const IPV4_LITERAL_RE = /^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$/;

function isBlockedIpv6Literal(hostname) {
  const h = hostname.toLowerCase();
  if (h === "::1" || h === "[::1]") return true;
  const stripped = h.replace(/^\[/, "").replace(/\]$/, "");
  // Unique local (fc00::/7) and link-local (fe80::/10) — checked on the
  // normalized (bracket-stripped) literal, not a full CIDR parse, since
  // these two prefixes only need a cheap prefix match.
  return /^f[cd][0-9a-f]{0,2}:/.test(stripped) || /^fe[89ab][0-9a-f]:/.test(stripped);
}

/**
 * @param {string} urlString
 * @returns {{ safe: boolean, reason: string|null }}
 */
export function isSafeUrl(urlString) {
  let parsed;
  try {
    parsed = new URL(urlString);
  } catch {
    return { safe: false, reason: "Not a valid URL" };
  }

  if (parsed.protocol !== "http:" && parsed.protocol !== "https:") {
    return { safe: false, reason: `Protocol must be http or https, got ${parsed.protocol}` };
  }

  const hostname = parsed.hostname.toLowerCase();
  if (BLOCKED_HOSTNAMES.has(hostname)) {
    return { safe: false, reason: `Hostname "${hostname}" is blocked` };
  }

  if (IPV4_LITERAL_RE.test(hostname)) {
    for (const [network, prefixLength] of BLOCKED_IPV4_RANGES) {
      if (isIpv4InRange(hostname, network, prefixLength)) {
        return { safe: false, reason: `IP ${hostname} is in a blocked private/internal range` };
      }
    }
  } else if (hostname.includes(":") || hostname.startsWith("[")) {
    // Bracketed or bare IPv6 literal.
    if (isBlockedIpv6Literal(hostname)) {
      return { safe: false, reason: `IPv6 address ${hostname} is a blocked loopback/private range` };
    }
  }

  // Non-standard ports: allow only the protocol's own default (i.e. no
  // explicit port at all — URL.port is "" when the port matches the
  // protocol default or was never specified).
  if (parsed.port !== "") {
    return { safe: false, reason: `Non-standard port ${parsed.port} is not allowed` };
  }

  return { safe: true, reason: null };
}

File: python/url_safety.py
"""
========================================================
URL_SAFETY.PY - SSRF validation, authoritative side (Rev 5 §9.6)
========================================================

Node's `services/urlSafety.js` does a cheap syntactic sanity check before
ever calling this service. This module is the AUTHORITATIVE check — per the
spec: "Python is the only side that sees the redirect chain." A hostname
that looks fine can still redirect to an internal IP mid-fetch, and DNS
rebinding means a hostname's resolved IP at validation time isn't
guaranteed to be its IP at fetch time either. So this module:

  1. Re-checks protocol/hostname/port the same way Node did (defense in
     depth — never trust a client-supplied "it's already validated").
  2. Resolves the hostname's actual IP address(es) and checks them against
     the private/loopback/link-local ranges using Python's `ipaddress`
     module (far more reliable than hand-rolled CIDR math).
  3. Manually follows the redirect chain (not via `requests`' automatic
     `allow_redirects=True`, which would fetch straight through to an
     unsafe hop before this code ever sees it) and re-validates every hop.
"""

import ipaddress
import socket
from urllib.parse import urlparse

import requests

BLOCKED_HOSTNAMES = {
    "localhost",
    "ip6-localhost",
    "ip6-loopback",
    "metadata.google.internal",
}

MAX_REDIRECTS = 5


class UnsafeUrlError(Exception):
    """Raised when a URL (or a hop in its redirect chain) fails SSRF validation."""


def _check_syntax(url):
    parsed = urlparse(url)
    if parsed.scheme not in ("http", "https"):
        raise UnsafeUrlError(f"Protocol must be http or https, got {parsed.scheme!r}")
    hostname = (parsed.hostname or "").lower()
    if not hostname:
        raise UnsafeUrlError("URL has no hostname")
    if hostname in BLOCKED_HOSTNAMES:
        raise UnsafeUrlError(f'Hostname "{hostname}" is blocked')
    # Non-standard port: urlparse only sets .port from what's literally in
    # the URL string, it does NOT normalize an explicit default (e.g.
    # "https://example.com:443/" still parses to port=443) — comparing
    # against None alone would wrongly reject a URL for writing out a port
    # that was already standard for its scheme.
    default_port = 443 if parsed.scheme == "https" else 80
    if parsed.port is not None and parsed.port != default_port:
        raise UnsafeUrlError(f"Non-standard port {parsed.port} is not allowed")
    return hostname


def _check_resolved_ips(hostname):
    """DNS-resolves hostname and rejects it if ANY resolved address is
    private/loopback/link-local/reserved. Checking all of them (not just
    the first) matters because a DNS response can list a safe address
    first and an internal one second — some resolvers/clients aren't
    guaranteed to always pick index 0."""
    try:
        addrinfo = socket.getaddrinfo(hostname, None)
    except socket.gaierror as exc:
        raise UnsafeUrlError(f"Could not resolve hostname {hostname!r}: {exc}") from exc

    for family, _, _, _, sockaddr in addrinfo:
        ip_str = sockaddr[0]
        ip = ipaddress.ip_address(ip_str)
        if ip.is_private or ip.is_loopback or ip.is_link_local or ip.is_reserved or ip.is_multicast:
            raise UnsafeUrlError(f"{hostname} resolves to blocked address {ip_str}")


def validate_redirect_chain(url, max_redirects=MAX_REDIRECTS):
    """Follows redirects manually, re-validating every hop's syntax AND
    resolved IP before following it. Returns the final, validated URL.
    Raises UnsafeUrlError the moment any hop fails validation."""
    current = url
    for _ in range(max_redirects + 1):
        hostname = _check_syntax(current)
        _check_resolved_ips(hostname)

        response = requests.head(current, allow_redirects=False, timeout=10)
        if response.status_code in (301, 302, 303, 307, 308):
            location = response.headers.get("Location")
            if not location:
                raise UnsafeUrlError(f"Redirect from {current} had no Location header")
            current = requests.compat.urljoin(current, location)
            continue
        return current

    raise UnsafeUrlError(f"Too many redirects (> {max_redirects}) starting from {url}")
